Source/Core/VideoCommon/Resources/MaterialResource.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2025 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <string_view>
+#include <variant>
+#include <vector>
+
+#include "Common/SmallVector.h"
+
+#include "VideoCommon/AbstractPipeline.h"
+#include "VideoCommon/Assets/MaterialAsset.h"
+#include "VideoCommon/Assets/TextureSamplerValue.h"
+#include "VideoCommon/Constants.h"
+#include "VideoCommon/GXPipelineTypes.h"
+#include "VideoCommon/RenderState.h"
+#include "VideoCommon/Resources/Resource.h"
+#include "VideoCommon/Resources/ShaderResource.h"
+#include "VideoCommon/Resources/TextureAndSamplerResource.h"
+
+namespace VideoCommon
+{
+class MaterialResource final : public Resource
+{
+public:
+  MaterialResource(Resource::ResourceContext resource_context, const GXPipelineUid& pipeline_uid);
+
+  struct TextureLikeReference
+  {
+    SamplerState sampler;
+    u32 sampler_index;
+    TextureSamplerValue::SamplerOrigin sampler_origin;
+    std::string_view texture_hash;
+    AbstractTexture* texture;
+  };
+
+  class Data
+  {
+  public:
+    AbstractPipeline* GetPipeline() const { return m_pipeline.get(); }
+    std::span<const u8> GetUniforms() const { return m_uniform_data; }
+    std::span<const TextureLikeReference> GetTextures() const { return m_texture_like_references; }
+    MaterialResource* GetNextMaterial() const { return m_next_material; }
+
+  private:
+    friend class MaterialResource;
+    std::unique_ptr<AbstractPipeline> m_pipeline = nullptr;
+    std::vector<u8> m_uniform_data;
+    std::shared_ptr<MaterialData> m_material_data = nullptr;
+    ShaderResource* m_shader_resource = nullptr;
+
+    // Variant for future expansion...
+    using TextureLikeResource = std::variant<TextureAndSamplerResource*>;
+    Common::SmallVector<TextureLikeResource, VideoCommon::MAX_PIXEL_SHADER_SAMPLERS>
+        m_texture_like_resources;
+    using TextureLikeData = std::variant<std::shared_ptr<TextureAndSamplerResource::Data>>;
+    Common::SmallVector<TextureLikeData, VideoCommon::MAX_PIXEL_SHADER_SAMPLERS>
+        m_texture_like_data;
+    Common::SmallVector<TextureLikeReference, VideoCommon::MAX_PIXEL_SHADER_SAMPLERS>
+        m_texture_like_references;
```

Comment:
```suggestion
    using TextureLikeResource = std::variant<TextureAndSamplerResource*>;
    Common::SmallVector<TextureLikeResource, VideoCommon::MAX_PIXEL_SHADER_SAMPLERS>
        m_texture_like_resources;
        
    using TextureLikeData = std::variant<std::shared_ptr<TextureAndSamplerResource::Data>>;
    Common::SmallVector<TextureLikeData, VideoCommon::MAX_PIXEL_SHADER_SAMPLERS>
        m_texture_like_data;
        
    Common::SmallVector<TextureLikeReference, VideoCommon::MAX_PIXEL_SHADER_SAMPLERS>
        m_texture_like_references;
```

I think some extra newlines could help readability.